File: composeApp/src/androidMain/kotlin/org/iremia/iremia/ui/journal/episode/EpisodeCaptureFlow.kt
package org.iremia.iremia.ui.journal.episode

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import org.iremia.iremia.data.garden.PlantResult
import org.iremia.iremia.domain.note.EntryType
import org.iremia.iremia.domain.note.EpisodeDraft
import org.iremia.iremia.ui.theme.IremiaColors
import java.util.Calendar

/** The stages of the capture flow. The type chooser comes first, then a branch. */
private enum class CaptureStage { TypeChooser, PanicIntensity, PanicContext, PanicReflection, Journal, Saved }

/** Insights goal shown on the confirmation screen. */
private const val INSIGHTS_GOAL = 30

/**
 * Self-contained capture flow. Opens on the entry-type chooser (panic vs journal),
 * then routes into either the existing panic wizard (intensity → context →
 * reflection) or the journal entry (guided prompts / free text). Both end on the
 * shared "saved" confirmation.
 *
 * The in-progress draft lives purely in UI state; [onSaveEpisode] persists the
 * finished [EpisodeDraft]. [onClose] backs out of the chooser, [onFinished]
 * dismisses after the confirmation screen.
 */
@Composable
fun EpisodeCaptureFlow(
    entryCount: Int,
    onClose: () -> Unit,
    onFinished: () -> Unit,
    onViewGarden: () -> Unit = onFinished,
    onSaveEpisode: (EpisodeDraft) -> Unit,
    // The result of the just-saved entry's plant attempt, for the saved screen's
    // conditional text (Block 3). Null until planting completes.
    plantResult: PlantResult? = null,
) {
    val finalEntryCount = remember { entryCount + 1 }
    var stage by rememberSaveable { mutableStateOf(CaptureStage.TypeChooser) }
    // The type + strength of the entry being saved, captured at save time so the
    // saved screen can show a matching impulse text (Block 3).
    var savedType by rememberSaveable { mutableStateOf(EntryType.PANIC) }
    var savedStrength by rememberSaveable { mutableStateOf<Int?>(null) }

    // --- Panic draft state (only used on the panic branch) ---
    val now = remember { Calendar.getInstance() }
    var selectedDateMillis by rememberSaveable { mutableStateOf(startOfDayUtc(now)) }
    var hour by rememberSaveable { mutableStateOf(now.get(Calendar.HOUR_OF_DAY)) }
    var minute by rememberSaveable { mutableStateOf(now.get(Calendar.MINUTE)) }
    var strength by rememberSaveable { mutableStateOf(6f) }
    val places = remember { mutableStateListOf<String>() }
    val activities = remember { mutableStateListOf<String>() }
    val bodySignals = remember { mutableStateListOf<String>() }
    var note by rememberSaveable { mutableStateOf("") }
    var moodBefore by rememberSaveable { mutableStateOf(-1) }
    var moodAfter by rememberSaveable { mutableStateOf(-1) }

    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(IremiaColors.White),
    ) {
        when (stage) {
            CaptureStage.TypeChooser -> EntryTypeChooserScreen(
                onBack = onClose,
                onPickPanic = { stage = CaptureStage.PanicIntensity },
                onPickJournal = { stage = CaptureStage.Journal },
            )

            CaptureStage.PanicIntensity -> EpisodeIntensityStep(
                dateMillis = selectedDateMillis,
                onDateChange = { selectedDateMillis = it },
                hour = hour,
                minute = minute,
                onTimeChange = { h, m -> hour = h; minute = m },
                strength = strength,
                onStrengthChange = { strength = it },
                onBack = { stage = CaptureStage.TypeChooser },
                onNext = { stage = CaptureStage.PanicContext },
                onSkip = { stage = CaptureStage.PanicContext },
            )

            CaptureStage.PanicContext -> EpisodeContextStep(
                places = places,
                onTogglePlace = { places.toggle(it) },
                activities = activities,
                onToggleActivity = { activities.toggle(it) },
                bodySignals = bodySignals,
                onToggleSignal = { bodySignals.toggle(it) },
                onBack = { stage = CaptureStage.PanicIntensity },
                onNext = { stage = CaptureStage.PanicReflection },
                onSkip = { stage = CaptureStage.PanicReflection },
            )

            CaptureStage.PanicReflection -> EpisodeReflectionStep(
                note = note,
                onNoteChange = { note = it },
                moodBefore = moodBefore,
                onMoodBefore = { moodBefore = it },
                moodAfter = moodAfter,
                onMoodAfter = { moodAfter = it },
                onBack = { stage = CaptureStage.PanicContext },
                onSave = {
                    savedType = EntryType.PANIC
                    savedStrength = strength.toInt()
                    onSaveEpisode(
                        EpisodeDraft(
                            content = note,
                            type = EntryType.PANIC,
                            strength = strength.toInt(),
                            places = places.toList(),
                            activities = activities.toList(),
                            bodySignals = bodySignals.toList(),
                            moodBefore = moodBefore.takeIf { it >= 0 },
                            moodAfter = moodAfter.takeIf { it >= 0 },
                            createdAt = combineDateTime(selectedDateMillis, hour, minute),
                        )
                    )
                    stage = CaptureStage.Saved
                },
            )

            CaptureStage.Journal -> JournalEntryScreen(
                onBack = { stage = CaptureStage.TypeChooser },
                onSave = { title, content ->
                    savedType = EntryType.JOURNAL
                    savedStrength = null
                    onSaveEpisode(
                        EpisodeDraft(
                            content = content,
                            type = EntryType.JOURNAL,
                            title = title.ifBlank { null },
                            // A journal entry lands "now"; no date/time picker.
                            createdAt = System.currentTimeMillis(),
                        )
                    )
                    stage = CaptureStage.Saved
                },
            )

            CaptureStage.Saved -> EpisodeSavedScreen(
                entryCount = finalEntryCount,
                goal = INSIGHTS_GOAL,
                entryType = savedType,
                strength = savedStrength,
                plantResult = plantResult,
                onInsights = onFinished,
                onHome = onFinished,
                onViewGarden = onViewGarden,
            )
        }
    }
}

/** Add [value] if absent, remove it if present (multi-select toggle). */
private fun <T> MutableList<T>.toggle(value: T) {
    if (!remove(value)) add(value)
}

/**
 * Start-of-day in UTC millis for the given calendar. Material's DatePicker works in
 * UTC, so we keep the selected date in UTC and only add the local time-of-day when
 * building the final timestamp.
 */
private fun startOfDayUtc(cal: Calendar): Long {
    val utc = Calendar.getInstance(java.util.TimeZone.getTimeZone("UTC"))
    utc.set(cal.get(Calendar.YEAR), cal.get(Calendar.MONTH), cal.get(Calendar.DAY_OF_MONTH), 0, 0, 0)
    utc.set(Calendar.MILLISECOND, 0)
    return utc.timeInMillis
}

/**
 * Combines a UTC start-of-day [dateMillis] (from the DatePicker) with a local
 * [hour]/[minute] into an epoch-millis timestamp in the device's zone, so the entry
 * lands on the chosen calendar day at the chosen time.
 */
private fun combineDateTime(dateMillis: Long, hour: Int, minute: Int): Long {
    val utc = Calendar.getInstance(java.util.TimeZone.getTimeZone("UTC")).apply { timeInMillis = dateMillis }
    val local = Calendar.getInstance().apply {
        set(utc.get(Calendar.YEAR), utc.get(Calendar.MONTH), utc.get(Calendar.DAY_OF_MONTH), hour, minute, 0)
        set(Calendar.MILLISECOND, 0)
    }
    return local.timeInMillis
}

File: iosApp/iosApp/Views/JournalViews/Prototype/Episode/EpisodeCaptureFlow.swift
import SwiftUI
import shared

// =============================================================================
// Capture flow — 1:1 translation of EpisodeCaptureFlow.kt.
// Opens on the entry-type chooser (panic vs journal), then branches.
// =============================================================================

/// The stages of the capture flow. The type chooser comes first, then a branch.
private enum CaptureStage {
    case typeChooser, panicIntensity, panicContext, panicReflection, journal, saved
}

/// Goal for the insights dataset.
private let insightsGoal = 30

/// The metadata the wizard collects for one entry before saving. Works for both
/// entry types: a panic entry fills the context/mood fields, a journal entry
/// usually only carries `content` and an optional `title`.
struct EpisodeDraftData {
    let content: String
    var type: String = "panic"
    var title: String? = nil
    let strength: Int?
    let places: [String]
    let activities: [String]
    let bodySignals: [String]
    let moodBefore: Int?
    let moodAfter: Int?
    /// When the entry happened (epoch millis).
    let createdAt: Int64

    /// A journal draft: text + optional title, no panic metadata.
    static func journal(content: String, title: String?, createdAt: Int64) -> EpisodeDraftData {
        EpisodeDraftData(
            content: content,
            type: "journal",
            title: title,
            strength: nil,
            places: [],
            activities: [],
            bodySignals: [],
            moodBefore: nil,
            moodAfter: nil,
            createdAt: createdAt
        )
    }

    /// Bridges to the shared `EpisodeDraft` KMP model for the controller.
    func toShared() -> EpisodeDraft {
        EpisodeDraft(
            content: content.trimmingCharacters(in: .whitespacesAndNewlines),
            type: type == "journal" ? .journal : .panic,
            title: title?.trimmingCharacters(in: .whitespacesAndNewlines),
            strength: strength.map { KotlinInt(int: Int32($0)) },
            places: places,
            activities: activities,
            bodySignals: bodySignals,
            moodBefore: moodBefore.map { KotlinInt(int: Int32($0)) },
            moodAfter: moodAfter.map { KotlinInt(int: Int32($0)) },
            createdAt: KotlinLong(longLong: createdAt)
        )
    }
}

/// Self-contained "Episode festhalten" wizard.
///
/// Holds the in-progress draft in UI state and saves the note to the database
/// on step 3. Returns the locked initial entry count + 1 on the confirmation screen.
struct EpisodeCaptureFlow: View {
    let entryCount: Int
    let onClose: () -> Void
    let onFinished: () -> Void
    let onViewGarden: () -> Void
    let onSaveEpisode: (EpisodeDraftData) -> Void
    /// Result of the just-saved entry's plant attempt, for the saved screen's
    /// conditional text (Block 3). Nil until planting completes.
    var plantResult: PlantResult?

    @State private var stage: CaptureStage = .typeChooser
    // The type + strength of the entry being saved, for the saved screen's impulse.
    @State private var savedIsJournal = false
    @State private var savedStrength: Int? = nil
    // The day the episode happened; defaults to today, past dates allowed.
    @State private var selectedDate: Date = Calendar.current.startOfDay(for: Date())
    @State private var hour: Int = Calendar.current.component(.hour, from: Date())
    @State private var minute: Int = Calendar.current.component(.minute, from: Date())
    @State private var strength: Float = 6
    @State private var places: [String] = []
    @State private var activities: [String] = []
    @State private var bodySignals: [String] = []
    @State private var note: String = ""
    @State private var moodBefore: Int = -1
    @State private var moodAfter: Int = -1

    @State private var finalEntryCount: Int

    init(entryCount: Int, onClose: @escaping () -> Void, onFinished: @escaping () -> Void, onViewGarden: @escaping () -> Void, onSaveEpisode: @escaping (EpisodeDraftData) -> Void, plantResult: PlantResult? = nil) {
        self.entryCount = entryCount
        self.onClose = onClose
        self.onFinished = onFinished
        self.onViewGarden = onViewGarden
        self.onSaveEpisode = onSaveEpisode
        self.plantResult = plantResult
        // Lock final entry count at init time to avoid double increments on DB sync
        _finalEntryCount = State(initialValue: entryCount + 1)
    }

    var body: some View {
        ZStack {
            IremiaColors.white.ignoresSafeArea()

            switch stage {
            case .typeChooser:
                EntryTypeChooserView(
                    onBack: onClose,
                    onPickPanic: { stage = .panicIntensity },
                    onPickJournal: { stage = .journal }
                )

            case .panicIntensity:
                EpisodeIntensityStepView(
                    selectedDate: $selectedDate,
                    hour: $hour,
                    minute: $minute,
                    strength: $strength,
                    onBack: { stage = .typeChooser },
                    onNext: { stage = .panicContext },
                    onSkip: { stage = .panicContext }
                )

            case .panicContext:
                EpisodeContextStepView(
                    places: $places,
                    activities: $activities,
                    bodySignals: $bodySignals,
                    onBack: { stage = .panicIntensity },
                    onNext: { stage = .panicReflection },
                    onSkip: { stage = .panicReflection }
                )

            case .panicReflection:
                EpisodeReflectionStepView(
                    note: $note,
                    moodBefore: $moodBefore,
                    moodAfter: $moodAfter,
                    onBack: { stage = .panicContext },
                    onSave: {
                        savedIsJournal = false
                        savedStrength = Int(strength)
                        onSaveEpisode(
                            EpisodeDraftData(
                                content: note,
                                type: "panic",
                                title: nil,
                                strength: Int(strength),
                                places: places,
                                activities: activities,
                                bodySignals: bodySignals,
                                moodBefore: moodBefore >= 0 ? moodBefore : nil,
                                moodAfter: moodAfter >= 0 ? moodAfter : nil,
                                createdAt: episodeTimestampMs(date: selectedDate, hour: hour, minute: minute)
                            )
                        )
                        stage = .saved
                    }
                )

            case .journal:
                JournalEntryView(
                    onBack: { stage = .typeChooser },
                    onSave: { title, content in
                        savedIsJournal = true
                        savedStrength = nil
                        let nowMs = Int64(Date().timeIntervalSince1970 * 1000)
                        onSaveEpisode(
                            EpisodeDraftData.journal(
                                content: content,
                                title: title.isEmpty ? nil : title,
                                createdAt: nowMs
                            )
                        )
                        stage = .saved
                    }
                )

            case .saved:
                EpisodeSavedScreenView(
                    entryCount: finalEntryCount,
                    goal: insightsGoal,
                    isJournal: savedIsJournal,
                    strength: savedStrength,
                    plantResult: plantResult,
                    onInsights: onFinished,
                    onHome: onFinished,
                    onViewGarden: onViewGarden
                )
            }
        }
    }
}

/// Combines a chosen day with an hour/minute into an epoch-millis timestamp in the
/// device's zone, so the entry lands on the chosen calendar day at the chosen time.
func episodeTimestampMs(date: Date, hour: Int, minute: Int) -> Int64 {
    var comps = Calendar.current.dateComponents([.year, .month, .day], from: date)
    comps.hour = hour
    comps.minute = minute
    comps.second = 0
    let combined = Calendar.current.date(from: comps) ?? date
    return Int64(combined.timeIntervalSince1970 * 1000)
}
